presto-main/src/main/java/com/facebook/presto/sql/planner/EqualityInference.java

```diff
@@ -355,55 +356,44 @@ public Builder addEquality(Expression expression)
 
         public Builder addEquality(Expression expression1, Expression expression2)
         {
-            checkArgument(!expression1.equals(expression2), "need to provide equality between different expressions");
             checkArgument(DeterminismEvaluator.isDeterministic(expression1), "Expression must be deterministic: " + expression1);
             checkArgument(DeterminismEvaluator.isDeterministic(expression2), "Expression must be deterministic: " + expression2);
 
-            Expression canonical1 = canonicalize(expression1);
-            Expression canonical2 = canonicalize(expression2);
-
-            if (!canonical1.equals(canonical2)) {
-                map.put(canonical1, canonical2);
-                reverseMap.put(canonical2, canonical1);
-            }
+            unionFindSet.findAndUnion(expression1, expression2);
             return this;
         }
 
-        private Expression canonicalize(Expression expression)
+        /**
+         * Performs one pass of generating more equivalences by rewriting sub-expressions in terms of known equivalences.
+         */
+        private void generateMoreEquivalences()
         {
-            while (map.containsKey(expression)) {
-                expression = map.get(expression);
+            Collection<Set<Expression>> equivalentClasses = unionFindSet.getEquivalentClasses();
+            Map<Expression, Set<Expression>> map = new HashMap<>();
+            for (Set<Expression> expressions : equivalentClasses) {
+                expressions.stream().forEach(expression -> map.put(expression, expressions));
             }
-            return expression;
-        }
 
-        private void collectEqualities(Expression expression, ImmutableSet.Builder<Expression> builder)
-        {
-            builder.add(expression);
-            for (Expression childExpression : reverseMap.get(expression)) {
-                collectEqualities(childExpression, builder);
+            for (Expression expression : map.keySet()) {
```

Comment:
This is a bit hard to follow. It may be a good idea to add some comments to illustrate with an example.